paderbox/utils/nested.py

```diff
@@ -400,3 +413,295 @@ def squeeze_nested(orig):
+    container[path[-1]] = value
+
+
+def nested_iter_items(container, *, iter_types=(dict, tuple, list)):
+    """
+
+    Args:
+        container:
+        iter_types:
+
+    Returns:
+
+    Examples:
+        >>> list(nested_iter_items({'a': {'b': {'c': [1, 2, 42], 'd': 'e'}}}))
+        [(('a', 'b', 'c', 0), 1), (('a', 'b', 'c', 1), 2), (('a', 'b', 'c', 2), 42), (('a', 'b', 'd'), 'e')]
+
+        This can be useful for example if you have a list of nested elements:
+        >>> c = [{'a': 1}, {'b': 2}, {'c': 2}]
+        >>> for idx, element in enumerate(c):
+        ...     for key, nested_element in element.items():
+        ...         print(idx, key, nested_element)
+        0 a 1
+        1 b 2
+        2 c 2
+
+        >>> for (idx, key), nested_element in nested_iter_items(c):
+        ...     print(idx, key, nested_element)
+        0 a 1
+        1 b 2
+        2 c 2
+    """
+    try:
+        items = container.items()
+    except AttributeError:
+        items = enumerate(container)
+    for key, value in items:
+        if isinstance(value, iter_types):
+            for key_, value_ in nested_iter_items(value):
+                yield (key, ) + key_, value_
+        else:
+            yield (key, ), value
+
+
+class FlatView:
+    def __init__(
+            self,
+            nested_container: Any,
+            *,
+            allow_partial_path: bool = False,
+            sep: str = '.',
+    ):
+        """
+        A view on a nested container that allows for access with paths. Useful
+        if the nested container contains non-dict objects so that a `flatten`
+        wouln't be reversible.
+
+        Note:
+            This view doesn't support `len` and `keys`, and thus is not a proper
+            `Mapping`, because these are not easily accessible
+            without traversing the whole nested structure.
+
+        Examples:
+            >>> d = {'a': 'b', 'c': {'d': {'e': 'f'}, 'g': [1, [2, 3], 4]}}
+            >>> v = FlatView(d)
+            >>> v['c.d.e']
+            'f'
+            >>> v.get('asdf', 'default')
+            'default'
+            >>> v['a'] = {'b': 'c'}
+            >>> v['a.b']
+            'c'
+            >>> d['a']['b']
+            'c'
+            >>> v[('c', 'g', 1, 1)]
+            3
+            >>> v = FlatView(d, allow_partial_path=True)
+            >>> v['a.b.c.d.e.f']
+            'c'
+            >>> list(v.items())
+            [(('a', 'b'), 'c'), (('c', 'd', 'e'), 'f'), (('c', 'g', 0), 1), (('c', 'g', 1, 0), 2), (('c', 'g', 1, 1), 3), (('c', 'g', 2), 4)]
+        """
+        self._container = nested_container
+        self.allow_partial_path = allow_partial_path
+        self.sep = sep
+
+    def get(self, item, default=..., *, sep=None):
+        if sep is None:
+            sep = self.sep
+        return get_by_path(
+            self._container, item, allow_early_stopping=self.allow_partial_path,
+            default=default, sep=sep,
+        )
+
+    def items(self):
```

Comment:
You could define keys (and values) as:

```python
def keys(self):
    return (k for k, _ in self.items())
```
(or using operator itemgetter)

Reply:
Usually, the `keys` function of a mapping returns a list or similar instead of a generator. Not sure if this can be confusing. We can also define `__len__` by `len(list(self.items()))`, but I don't want to run through the whole nested container for this

Reply:
Partial. I think in python2 keys, values and items yielded lists, but in python 3 they started to return special types, and now you mostly cast them to a list or tuple to work with them.

My motivation is more: When  `items` is defined, I expect that `keys` and `values` exists.
I agree, that all of them aren't so well-defined for this type.

Reply:
I agree with you, I also expect `keys` and `values` to exist when `items` exists, but the definition of `keys` and `items` is different from a typical `Mapping`.  But since we can't subclass `Mapping` because we can't define `__len__`, I think it's fine.

Reply:
Following your arguments, keys is not well-defined, hence we should not provide `items`?
With the definition of items, you fixed everything. The only problem is, that execution time of `__len__` is not constant (`keys`, `values` and `items` is constant, because the computation happens, when you iterate.).

Reply:
Yes, that's right. So you say, we should also provide `__len__`?

Another problem is that `items` yields the leaves and not all possible keys you can index the `FlatView` with.

Reply:
Here, the things that I think have arguments (I am fine with the first three, with a small preference of the first):
 - Provide None
 - Provide `items`, `values`, `keys`
 - Provide `items`, `values`, `keys`, `__len__`
 - Provide `values` (Difficult, because values should not contain duplicates)

My argumentation is that I cannot find an argument, to only implement `items`.
It is implemented as a generator, and `values` and `keys` can have the same implementation (My example was a generator comprehension) and `operator.itemgetter` is also lazy.

An implementation of `__len__` is optional, because it is relative expensive, while it shouldn't be the case for an implementation of `__len__` (We can compensate it with caching).

My small preference to provide non, is the fact, that I don't see the use case and the non-unique definition (Once we implement it, a change of the definition is a breaking change).

Reply:
> An implementation of __len__ is optional, because it is relative expensive, while it shouldn't be the case for an implementation of __len__ (We can compensate it with caching).

Unfortunately we can't use caching: we can't detect changes in the wrapped container, so we don't know when to reset the cache

My argument for `items` is that I implemented `nested_iter_items` and had some use-cases in the past where I needed something like this. I would be confused if `FlatView` wouldn't support the same functionality in some way and `items` is the most obvious place for it.

We could also implement an `items` which iterates over everything and an `leaf_items` that only iterates over the leaves.